python/thunder/imgprocessing/transformation.py

```diff
@@ -8,6 +9,168 @@ def apply(self, im):
         raise NotImplementedError
 
 
+class DifferentiableTransformation(Transformation):
+    """
+    Differentiable transformations must have methods to compute the Jacobian and update parameters. These are used by
+    iterative alignment techniques like Lucas-Kanade and RASL.
+    """
+    def jacobian(self, imageGradients, imageGrid):
+        """Compute gradient of transformation output with respect to parameters.
+
+        Parameters
+        ----------
+        imageGradients : list of volumes, one for each dimension
+            gradient of volume in each dimension
+        imageGrid : list of arrays, one for each dimension
+            each array contains the coordinates along that dimension for each voxel
+
+        Returns
+        -------
+        ordered list of arrays containing the jacobian for each parameter
+        """
+        raise NotImplementedError
+
+    def updateParams(self, deltaParams):
+        """Update each parameter in the transformation.
+
+        Parameters
+        ----------
+        deltaParams : array, shape (nparams,)
+            ordered changes in parameters to be applied. The ordering here will match
+            the ordering of the parameters returned by jacobian.
+        """
+        raise NotImplementedError
+
+
+class AffineTransformation(DifferentiableTransformation):
+    """ Affine transformations are differentiable and can be represented as a matrix."""
+    def matrix(self):
+        raise NotImplementedError
+
+    def apply(self, vol, grid=None, order=1):
+        if grid is None:
+            center = getattr(self, 'center', None)
+            grid = GridTransformer(vol.shape, center=center)
+        A = self.matrix()
+        return grid.transform_vol(vol, A, order=order)
+
+
+class TranslationTransformation(AffineTransformation):
```

Comment:
Is it a concern that we have effectively duplicated here the `Displacement` transformation method defined below? The implementation (with the grid transformer) is very different, but conceptually they are very similar right? Except that this can also be sub-pixel? Maybe then it's a just a semantic problem that `Displacement` and `Affine` sound more similar than their implementation reflects.